### ISSUE TITLE
psyllid with fpa receiver stops after a while

### ISSUE THREAD

**Issue by cclaessens** [project8/psyllid#17]
After activating and about a minute psyllid stops.

[fpa_receiver_stops_after_a_while.txt](https://github.com/project8/psyllid/files/586197/fpa_receiver_stops_after_a_while.txt)

**Comment by nsoblath** [comment]
I've replicated this error.  In my tests it always died between the 4,247,000th packet and the 4,248 000th packet:

```
21:43:04 [ INFO] /core/_buffer.hh(195): write stream <0x7fffec000e20> processed <4247000> requests

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff13e3700 (LWP 6943)]
__memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:35
35	../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S: No such file or directory.
```

**Comment by nsoblath** [comment]
I just tried to replicate it again, and the problem happened here:

```
17:28:18 [ INFO] /core/_buffer.hh(226): write stream <0x7fffec0008c0> processed <2120000> requests
17:28:18 [ INFO] /core/_buffer.hh(226): write stream <0x7fffec0008c0> processed <2121000> requests
17:28:18 [ INFO] /core/_buffer.hh(226): write stream <0x7fffec0008c0> processed <2122000> requests
17:28:18 [ INFO] /core/_buffer.hh(226): write stream <0x7fffec0008c0> processed <2123000> requests

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff13f5700 (LWP 15758)]
__memcpy_sse2_unaligned () at ../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S:35
35	../sysdeps/x86_64/multiarch/memcpy-sse2-unaligned.S: No such file or directory.
```

Suspicious thing to note: that appears to be half-way to where it crashed for me before.

**Comment by nsoblath** [comment]
This is a problem with the `memcpy` call that's currently at line 384 of packet_receiver_fpa.cc:

```c++
        ::memcpy( reinterpret_cast< void* >( t_mem_block->block() ),
                  reinterpret_cast< void* >( (char*)t_udp_hdr + t_udp_hdr_len ),
                  t_udp_data_len );
```

A segmentation fault indicates a problem with either the memory that's being read from or the memory that's being written to.  Here's an example of what those pointers should look like, printed when the logger verbosity is set to `trace`:

```
14:16:01 [TRACE] _receiver_fpa.cc(394): UDP mem addresses (packet/header, data): 0x7f368d7ef564, 0x7f368d7ef56c
14:16:01 [TRACE] _receiver_fpa.cc(399): Packet received (8224 bytes); block address is 0x7f34dc55f010
```

To determine which pointer is at fault, I'll catch the SEGSEGV signal that's thrown and print out the pointer values when the segfault occurs:

```
14:18:41 [ INFO] /core/_buffer.hh(226): write stream <0x7f67c80008c0> processed <2123000> requests
14:18:41 [ERROR] _receiver_fpa.cc(52): Received signal 11
14:18:41 [ERROR] _receiver_fpa.cc(53): Destination (block address): 0
14:18:41 [ERROR] _receiver_fpa.cc(54): Source (udphdr + hdr_len): 0x7f67ae62ebe4 + 8 = 0x7f67ae62ebec
14:18:41 [ERROR] _receiver_fpa.cc(55): Data len: 8224
```

So it's clearly the destination pointer that's the problem.

Digging into what was going on with the output stream of `packet_receiver_fpa`, I determined that at some point after block 2123000, but before 2124000, the `block` pointer in `memory_block` was being reset to 0.    That led me to pretty quickly realize that in `memory_block::resize()` the value of f_n_bytes was never being reset to the new size.  I'm not sure exactly how that led the `block` pointer being set to 0 or at least returning 0, but that's what happened.

After fixing the `memory_block::resize()` function, psyllid was able to run stably past where it was crashing before.  So far it's up to 42M packets processed, and it's still going.

This is now fixed as of commit d3d7a62d970a6f9745c0cbd4ed83e84d93ffdb67.
